src/stdlib/builtins.rs

```diff
@@ -51,17 +51,35 @@ fn assert_eq_fn<B: Backend>(
 
     // they are both of type field
     if !matches!(lhs_info.typ, Some(TyKind::Field { .. })) {
-        panic!(
-            "the lhs of assert_eq must be of type Field or BigInt. It was of type {:?}",
-            lhs_info.typ
-        );
+        let lhs = lhs_info.typ.clone().ok_or_else(|| {
+            Error::new(
+                "constraint-generation",
+                ErrorKind::UnexpectedError("No type info for lhs of assertion"),
+                span,
+            )
+        })?;
+
+        Err(Error::new(
+            "constraint-generation",
+            ErrorKind::AssertTypeMismatch("rhs", lhs),
+            span,
+        ))?
     }
 
     if !matches!(rhs_info.typ, Some(TyKind::Field { .. })) {
-        panic!(
-            "the rhs of assert_eq must be of type Field or BigInt. It was of type {:?}",
-            rhs_info.typ
-        );
+        let rhs = rhs_info.typ.clone().ok_or_else(|| {
+            Error::new(
+                "constraint-generation",
+                ErrorKind::UnexpectedError("No type info for rhs of assertion"),
+                span,
+            )
+        })?;
+
+        Err(Error::new(
+            "constraint-generation",
+            ErrorKind::AssertTypeMismatch("rhs", rhs),
+            span,
```

Comment:
I think UnexpectedError is still good here, as I don't think we should run into these things (the type checker will catch that before we hit this place)

Reply:
The main issue that I ran into here is that I wanted to also include `res_info.typ` in the error message just like in the panic version. format!()` returns `String` and `UnexpectedError` requires `&'static str` as its tagged field. Borrowing the `String` does not last for `'static` unless I leak the allocation. In order to get around this, I created this new error variant. The alternative was a slightly more opaque error message. I'm not sure if this is the right thing per se, but worth discussing briefly